### DISCUSSION TITLE
And primitive/parser

### DISCUSSION THREAD

**Discussion by SephVelut** [zesterer/chumsky#135] [Ideas]
I need to match an ident that begins with an uppercase. I need to the result to be in a complete form, including all the tokens matched in one string. Closest I can get to doing this

```rust
    let uppercase_char = filter(|c: &char| c.is_uppercase());

    let spec_name = just::<_, _, Simple<char>>('#')
        .ignore_then(uppercase_char.ignored().rewind().then(text::ident())).parse("#Treehouse");
```

This still results in `Ok(((), "Treehouse"))`. The tuple, having a first item of `()`, is undesirable. I could probably also use `to` to convert it to the desired result. Ideally, it should only be `Ok("Treehouse")`. Seems like this needs a `and` primitive that allows for chaining multiple methods that match tokens in sequence, each with their own predicate, then collates the result as a single output.

`just("a").and(just("b")).and(just("c")).parse("abc")` should give `Ok("abc")`.

Currently, `just('a').then(just('b')).then(just('c')).parse("abc")` gives `Ok((('a', 'b'), 'c'))`. The tuples here are highly undesirable for me.

**Comment by zesterer** [comment]
I think you should just be able to use `ignore_then` twice:

```rs
let uppercase_char = filter(|c: &char| c.is_uppercase());

let spec_name = just::<_, _, Simple<char>>('#')
    .ignore_then(uppercase_char.ignored().rewind())
    .ignore_then(text::ident()).parse("#Treehouse");
```

**Comment by zesterer** [comment]
Regarding the `and` combinator, you might find that [`chain`](https://docs.rs/chumsky/latest/chumsky/trait.Parser.html#method.chain) does effectively the same thing, but in a more generalised way.
